[docs] Add roadmap

abiogenesis-now opened the pull request:
Intended to address https://github.com/ksonnet/ksonnet/issues/207 (at least partially).

Unlike with [Contour's design docs](https://github.com/heptio/contour/tree/master/design), I opted to nest these inside of `docs/`. I think that this sort of organization is actually simpler because it's one less top-level directory to check, but if anyone feels strongly about that I'm happy to discuss and change.

This also might not be the proper structure/format for a roadmap (it seems to overlap a fair amount with stuff in the [issues](https://github.com/ksonnet/ksonnet/issues)?), but I'm hoping that even if we don't merge this in, it proves helpful in the planning/prioritization discussion tomorrow.

cc @jessicayuen @hausdorff @bryanl 

Signed-off-by: Jessica Yao <jessica@heptio.com>

abiogenesis-now commented:
when i get a chance today, will review/edit this and tag folks

abiogenesis-now commented:
@bryanl @jessicayuen @hausdorff PTAL when you get a chance!

If you think the roadmap descriptions are too long/redundant with what's on the project board, **feel free to ask me to cut things out**. I'm fine with removing the descriptions and just linking out in the future if that's faster.

A lot of the features described require their own design docs----it's fine to draft these in Google docs, but when they're done, do you mind also dropping them into the `docs/design` folder (post-merging of this PR)? This is what the Contour repo does. Thank you!

mumoshu commented:
(Mentioning only who seems active recently not to bother you too much)
Hi @abiogenesis-now @bryanl, thanks again for maintaining the great project 👍 

Would you mind including #255 into the roadmap, probably marked as "contribution welcomed" if it makes sense?
I'm willing to contribute it myself. However, I am not yet sure if it really (seem to) help solving the problem this project aims without comments from you core maintainers. Thanks.

mumoshu commented on docs/design/roadmap.md:
```diff
@@ -0,0 +1,168 @@
+* [Strategic, key-based merge](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#use-a-strategic-merge-patch-to-update-a-deployment) (`kubectl patch`)
+* [Three-way merge](https://kubernetes.io/docs/tutorials/object-management-kubectl/declarative-object-management-configuration/#how-apply-calculates-differences-and-merges-changes) (`kubectl apply`).
+
+**`ks` will likely adopt `kubectl`'s merge strategies to be in line with what users are familiar with**, and work with upstream `kubectl` to remove other unintuitive behaviors (documented in some [integration tests](https://github.com/hausdorff/ksonnet/commit/8e8522938fe0f940cd5af6f19bc1fa47bf24bfc6)). This overlaps with the work of the [Declarative App Def WG](https://github.com/kubernetes/commu ...
```
FWIW [kubectl apply --prune](https://github.com/kubernetes/kubernetes/issues/19805) and its [implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply.go#L229) would help.

mumoshu commented on docs/design/roadmap.md:
```diff
@@ -0,0 +1,168 @@
+* Deleting a component in your ksonnet app is actually reflected server-side
+* It's generally easier for users to reason through what will run on their cluster
+
+However, it also leads to bugs where `ks apply` clobbers fields that ought to have been retained. This means that users get a different `nodePort` each time they `ks apply` a `ClusterIP` Service, and that changes from horizontal pod autoscaling are lost.
```
Just curious but is `ks apply` with the AnnotationGcStrategy meant for "recreating" every object managed by ksonnet, hence `nodePort` changes each time you run `ks apply` as noted here, according to code [here](https://github.com/ksonnet/ksonnet/blob/10bbf40f4383b27f72c5fa1aeff24e339729f010/pkg/kubecfg/apply.go#L88) and [there](https://github.com/ksonnet/ksonnet/blob/10bbf40f4383b27f72c5fa1aeff24e339729f010/cmd/apply.go#L119-L122)?

bryanl replied:
The goal here is to maintain items in a declarative fashion. If they don't exist, create them.  If they do exist and aren't needed any more, delete them.

bryanl commented on docs/design/roadmap.md:
```diff
@@ -0,0 +1,168 @@
+
+**An associated command (e.g. `ks update`) should allow users to download the latest versions of a dependency.**
+
+## Provide access to "ambient parameters"
```
The meaning of ambient parameters isn't immediately obvious. Is there a better way to phrase this?

bryanl commented on docs/design/roadmap.md:
```diff
@@ -0,0 +1,168 @@
+
+**Ideally, the fields that ksonnet modifies are tracked in a way that makes it possible to hide other, non-essential fields during `ks diff`.** This work may tie into the [changes to `ks apply` semantics](#establish-clear-ks-apply-semantics). Proposed solutions include an annotation that lists ksonnet-specific fields (so that other can be filtered), but more discussion is necessary.
+
+## Surface errors in app files (`ks lint`)
```
Could you reword this? I am going to suggest, "Show errors in app files"

abiogenesis-now commented:
@bryanl - addressed your comments, let me know if you have any other feedback! 

I incorporated a description for secrets (per @mumoshu 's request), but didn't include too much detail because I think that issue needs to be discussed further. If we decide that encryption is not in the scope of the ksonnet tool itself, we can still provide documentation/tutorials to support integration with other tools (e.g. sealed secrets, with caveats).
